src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -55,46 +55,121 @@
-    throws ServletException, IOException {
-    String stringToMatch = request
-      .getParameter("searchterm")
-      .replaceAll("[^a-zA-Z0-9]+", " ")
-      .trim();
-    request.setAttribute("dataSearched", stringToMatch);
+    throws ServletException, IOException, NullPointerException {
+    List<Recipe> recipesListToReturn = new ArrayList<>();
+    Integer timeValue;
+    String stringToMatch = getParameter("searchterm", request);
+    String difficulty = getParameter("difficulty", request);
+    String time = getParameter("time", request);
+    HashMap<String, String> difficultiesMap = createMapOfDifficulties();
+    HashMap<Integer, String> timeIntervalsMap = createMapOfTimeIntervals();
 
-    List<Recipe> recipesListToReturn = recipesMatching(
-      request,
-      response,
-      stringToMatch
+    request.setAttribute("dataSearched", stringToMatch);
+    request.setAttribute("difficultySearchedKey", difficulty);
+    request.setAttribute("timeSearchedKey", time);
+    request.setAttribute(
+      "difficultySearchedValue",
+      difficultiesMap.get(difficulty)
     );
+    request.setAttribute("difficultyList", difficultiesMap);
+    request.setAttribute("cookingTimeList", timeIntervalsMap);
 
+    stringToMatch = trimAndRemoveCommas(stringToMatch);
+    difficulty = trimAndRemoveCommas(difficulty);
+    try {
+      timeValue = Integer.valueOf(trimAndRemoveCommas(time));
+    } catch (NumberFormatException e) {
+      timeValue = 0;
+    }
+    request.setAttribute("timeSearchedValue", timeIntervalsMap.get(timeValue));
+
+    try {
+      recipesListToReturn =
+        recipesMatching(
+          request,
+          response,
+          stringToMatch,
+          difficulty,
+          timeValue
+        );
+    } catch (SearchQueryException e) {
+      request.getRequestDispatcher("/search.jsp").forward(request, response);
+    }
     request.setAttribute("recipesList", recipesListToReturn);
     request.getRequestDispatcher("/search.jsp").forward(request, response);
   }
 
+  /**
+   * Creates a map of values for the dropdown select for difficulties.
+   */
+  private HashMap<String, String> createMapOfDifficulties() {
+    HashMap<String, String> options = new HashMap<String, String>();
+    options.put("hard", "Hard");
+    options.put("medium", "Medium");
+    options.put("easy", "Easy");
+    return options;
+  }
+
+  /**
+   * Creates a map of values for the dropdown select for difficulties.
+   */
+  private HashMap<Integer, String> createMapOfTimeIntervals() {
+    HashMap<Integer, String> options = new HashMap<Integer, String>();
+    options.put(120, "Less than 2h");
+    options.put(60, "Less than 1h");
+    options.put(30, "Less than 30 minutes");
+    return options;
+  }
+
+  /**
+   * Replace and return the string with no commas and no multiple consecutive or trailing spaces.
+   */
+  private String trimAndRemoveCommas(String stringToRemove) {
```

Comment:
Why the parameter is called stringToRemove? You're not removing the string. I think the function should be called "sanitize" and the parameter could be only "text" or "string".